Harvest LDD errors don't provide enough context to know what is going on

### Checked for duplicates

Yes - I've already checked

### 🐛 Describe the bug

When I tried to load the MRO/SHARAD bundle, I get an `Could not update LDD error and I don't know what this means.

### 🕵️ Expected behavior

I expected to not receive an error or receive an error I can actually interpret

### 📜 To Reproduce

1. try to load https://pds-geosciences.wustl.edu/mro/mro-m-sharad-5-3d-v1/mrosh_3001/
2. see error in log file ([sharadLog.txt](https://github.com/user-attachments/files/22913890/sharadLog.txt))


### 🖥 Environment Info

Windows OS


### 📚 Version of Software Used

v4.0.7

### 🩺 Test Data / Additional context

https://pds-geosciences.wustl.edu/mro/mro-m-sharad-5-3d-v1/mrosh_3001/

### 🦄 Related requirements

🦄 #xyz

### ⚙️ Engineering Details

_No response_

### 🎉 Integration & Test

_No response_

@tloubrieu-jpl it looks like the errors looked like this:

```
2025-08-13 13:00:34 [ERROR] (SchemaUpdater.java:updateSchema:86) Could not update LDD. The system cannot find the path specified
```

--- a/.github/workflows/secrets-detection.yaml
+++ b/.github/workflows/secrets-detection.yaml
@@ -17,7 +17,7 @@ jobs:
             -
                 name: Install necessary packages
                 run: |
-                    pip install git+https://github.com/NASA-AMMOS/slim-detect-secrets.git@exp
+                    pip install detect-secrets==1.5.0
                     pip install jq
               
             -

--- a/.github/workflows/secrets-detection.yaml
+++ b/.github/workflows/secrets-detection.yaml
@@ -44,7 +44,7 @@ jobs:
                     cp .secrets.baseline .secrets.new
 
                     # find the secrets in the repository
-                    detect-secrets scan --disable-plugin AbsolutePathDetectorExperimental --baseline .secrets.new \
+                    detect-secrets scan --baseline .secrets.new \
                         --exclude-files '\.secrets..*' \
                         --exclude-files '\.pre-commit-config\.yaml' \
                         --exclude-files '\.git.*' \

--- a/.secrets.baseline
+++ b/.secrets.baseline
@@ -1,14 +1,11 @@
 {
-  "version": "1.4.0",
+  "version": "1.5.0",
   "plugins_used": [
     {
       "name": "ArtifactoryDetector"
     },
     {
       "name": "AWSKeyDetector"
-    },
-    {
-      "name": "AWSSensitiveInfoDetectorExperimental"
     },
     {
       "name": "AzureStorageKeyDetector"
@@ -27,10 +24,10 @@
       "name": "DiscordBotTokenDetector"
     },
     {
-      "name": "EmailAddressDetector"
+      "name": "GitHubTokenDetector"
     },
     {
-      "name": "GitHubTokenDetector"
+      "name": "GitLabTokenDetector"
     },
     {
       "name": "HexHighEntropyString",
@@ -59,7 +56,13 @@
       "name": "NpmDetector"
     },
     {
+      "name": "OpenAIDetector"
+    },
+    {
       "name": "PrivateKeyDetector"
+    },
+    {
+      "name": "PypiTokenDetector"
     },
     {
       "name": "SendGridDetector"
@@ -75,6 +78,9 @@
     },
     {
       "name": "StripeDetector"
+    },
+    {
+      "name": "TelegramBotTokenDetector"
     },
     {
       "name": "TwilioKeyDetector"
@@ -122,23 +128,30 @@
     {
       "path": "detect_secrets.filters.regex.should_exclude_file",
       "pattern": [
-        "\\.secrets..*",
         "\\.git.*",
         "\\.pre-commit-config\\.yaml",
-        "target"
+        "\\.secrets..*",
+        "target",
+        ".*\\.tfstate",
+        ".*\\.tfstate.*",
+        ".*\\.tfvars",
+        ".*\\.terraform",
+        ".*\\.terraformrc",
+        ".*\\terraform.rc"
       ]
     }
   ],
   "results": {
-    "pom.xml": [
+    "CLAUDE.md": [
       {
-        "type": "Email Address",
-        "filename": "pom.xml",
-        "hashed_secret": "fac2dea9e49a83a2d6ee38c580d1e5358b45efa5",
+        "type": "Secret Keyword",
+        "filename": "CLAUDE.md",
+        "hashed_secret": "e5413f44bc692c1cbddadb8d29aa82f7ebbdb75e",
         "is_verified": false,
-        "line_number": 189
+        "line_number": 115,
+        "is_secret": false
       }
     ]
   },
-  "generated_at": "2025-06-19T20:00:34Z"
+  "generated_at": "2026-04-22T18:50:50Z"
 }

--- a/CLAUDE.md
+++ b/CLAUDE.md
@@ -1,0 +1,132 @@
+# CLAUDE.md
+
+This file provides guidance to Claude Code (claude.ai/code) when working with code in this repository.
+
+## Project Overview
+
+This is a Java library providing common functionality for interacting with the PDS (Planetary Data System) Registry and Elasticsearch/OpenSearch. It serves as a shared dependency for registry components like Harvest and Registry Manager.
+
+**Key characteristics:**
+- Maven-based Java project (Java 17)
+- Version: 2.3.0
+- Supports both Elasticsearch REST client (SDK v1) and OpenSearch Java SDK (SDK v2)
+- Provides abstraction layer for multiple connection types: direct, Cognito-authenticated, and EC2
+
+## Build Commands
+
+```bash
+# Build the project
+mvn clean install
+
+# Build without tests
+mvn clean install -DskipTests
+
+# Run tests
+mvn test
+
+# Run a single test
+mvn test -Dtest=TestClassName
+
+# Package the JAR
+mvn package
+
+# Create source JAR (for debugging)
+mvn source:jar
+
+# Clean build artifacts
+mvn clean
+```
+
+## Code Architecture
+
+### Connection Architecture
+
+The library implements a dual-SDK architecture to support both Elasticsearch REST client and OpenSearch SDK:
+
+**Connection Factory Pattern:**
+- `ConnectionFactory` interface - defines contract for creating connections
+- `EstablishConnectionFactory` - factory entry point that routes to appropriate SDK implementation
+- `UseOpensearchSDK1` - implements connections using Elasticsearch REST client (older, direct connections)
+- `UseOpensearchSDK2` - implements connections using OpenSearch Java SDK (supports AWS Cognito, EC2, serverless)
+
+**Connection Types:**
+1. **Direct connections** - Direct HTTP/HTTPS to OpenSearch/Elasticsearch (supports both SDK v1 and v2)
+2. **Cognito connections** - AWS Cognito authentication for serverless OpenSearch (SDK v2 only)
+3. **EC2 connections** - EC2 instance role credentials (SDK v2 only)
+
+Connection configurations are stored as XML files in `src/main/resources/connections/` with separate subdirectories for `cognito/` and `direct/` connection types.
+
+### Core Packages
+
+**`gov.nasa.pds.registry.common.connection`** - Connection management and authentication
+- `aws/` - AWS OpenSearch SDK v2 implementations (serverless support)
+- `es/` - Elasticsearch REST client SDK v1 implementations
+- `config/` - JAXB-generated classes for parsing connection XML configurations
+
+**`gov.nasa.pds.registry.common.meta`** - Metadata extraction from PDS labels
+- Extractors for different product types: Bundle, Collection, File, etc.
+- `Metadata` class represents extracted metadata with LID, LIDVID, internal references, and fields
+- `MetadataNormalizer` handles field normalization for Elasticsearch indexing
+
+**`gov.nasa.pds.registry.common.dd`** - Data Dictionary handling
+- Maps PDS data types to Elasticsearch data types
+- Parses LDD (Local Data Dictionary) and PDD (Planetary Data Dictionary) files
+- `DDRecord` represents data dictionary entries for indexing
+
+**`gov.nasa.pds.registry.common.es`** - Elasticsearch/OpenSearch operations
+- `dao/` - Data Access Objects for products, schema, and data dictionaries
+- `service/` - Higher-level services like ProductService, CollectionInventoryWriter
+
+**`gov.nasa.pds.registry.common.util`** - Utility classes for file handling, date parsing, XML/JSON processing
+
+### Request/Response Abstraction
+
+The library provides a unified `Request` and `Response` interface that abstracts over both SDKs:
+- `RestClient` interface defines operations (bulk, search, count, get, delete, etc.)
+- SDK-specific implementations in `connection/aws/` and `connection/es/` wrap native SDK objects
+- This allows client code (Harvest, Registry Manager) to work with either SDK transparently
+
+### Known Registries
+
+`KnownRegistryConnections` discovers and lists all available connection configurations by scanning resources. It initializes a custom URL protocol handler (`app://`) to load connection configs from classpath resources.
+
+## Development Notes
+
+**Testing:**
+- Tests are in `src/test/java/`
+- Test classes organized by functionality: `dao/`, `meta/`, `tt/` (technical tests), `xml/`
+- Some tests may require a running Elasticsearch/OpenSearch instance
+
+**Connection configuration:**
+- XML schema defines connection types and their properties
+- SDK version is specified in XML: `<server_url sdk="1">` or `sdk="2"`
+- Default SDK for direct connections is 1 (Elasticsearch REST client)
+- Cognito and EC2 connections always use SDK 2
+
+**Pre-commit hooks:**
+- Secret detection using `slim-detect-secrets`
+- Run `pre-commit install` to enable hooks locally
+- Baseline file: `.secrets.baseline`
+
+**CI/CD:**
+- Unstable builds (main branch): `.github/workflows/unstable-cicd.yaml`
+- Stable builds (releases): `.github/workflows/stable-cicd.yaml`
+- Builds publish to Maven Central via Sonatype Central Portal
+- Requires secrets: `CENTRAL_REPOSITORY_USERNAME`, `CENTRAL_REPOSITORY_TOKEN`, `CODE_SIGNING_KEY`
+
+## Important Patterns
+
+**Metadata field naming:**
+- Uses PDS4 hierarchical naming: `namespace:ClassName/namespace:attribute_name`
+- Namespace separator: `:` (defined in `MetaConstants.NS_SEPARATOR`)
+- Attribute separator: `/` (defined in `MetaConstants.ATTR_SEPARATOR`)
+- Example: `pds:Product_Observational/pds:title`
+
+**Error handling:**
+- `ResponseException` wraps SDK-specific exceptions
+- SDK1 uses `ResponseExceptionWrapper` to wrap Elasticsearch exceptions
+- SDK2 uses AWS SDK exceptions directly
+
+**Resource management:**
+- `RestClient` implements `Closeable` - always use try-with-resources
+- `CloseUtils` provides helper methods for safely closing resources

--- a/src/main/java/gov/nasa/pds/registry/common/dd/LddException.java
+++ b/src/main/java/gov/nasa/pds/registry/common/dd/LddException.java
@@ -1,0 +1,26 @@
+package gov.nasa.pds.registry.common.dd;
+
+/**
+ * Thrown when a Local Data Dictionary (LDD) cannot be downloaded or loaded
+ * into the registry data dictionary index.
+ */
+@SuppressWarnings("serial")
+public class LddException extends Exception {
+
+  /**
+   * Constructor
+   * @param message description of the failure
+   */
+  public LddException(String message) {
+    super(message);
+  }
+
+  /**
+   * Constructor
+   * @param message description of the failure
+   * @param cause the underlying exception
+   */
+  public LddException(String message, Throwable cause) {
+    super(message, cause);
+  }
+}

--- a/src/main/java/gov/nasa/pds/registry/common/es/service/SchemaUpdater.java
+++ b/src/main/java/gov/nasa/pds/registry/common/es/service/SchemaUpdater.java
@@ -2,6 +2,7 @@ package gov.nasa.pds.registry.common.es.service;
 
 
 import java.io.File;
+import java.io.IOException;
 import java.util.List;
 import java.util.Map;
 import java.util.Set;
@@ -120,7 +121,9 @@ public class SchemaUpdater {
     LddVersions lddInfo;
     try {
       lddInfo = ddDao.getLddInfo(prefix);
-    } catch (Exception ex) {
+    } catch (RuntimeException ex) {
+      throw ex;
+    } catch (IOException ex) {
       throw new LddException("Failed to query registry for existing LDD info for namespace '"
           + prefix + "': " + ExceptionUtils.getMessage(ex), ex);
     }
@@ -134,7 +137,7 @@ public class SchemaUpdater {
     File lddFile;
     try {
       lddFile = File.createTempFile("LDD-", ".JSON");
-    } catch (Exception ex) {
+    } catch (IOException ex) {
       throw new LddException("Failed to create temp file for LDD download for namespace '"
           + prefix + "': " + ExceptionUtils.getMessage(ex), ex);
     }
@@ -143,17 +146,27 @@ public class SchemaUpdater {
       if (fileDownloader.download(jsonUrl, lddFile)) {
         lddLoader.load(lddFile, schemaFileName, prefix);
       }
+    } catch (RuntimeException ex) {
+      throw ex;
+    } catch (InterruptedException ex) {
+      Thread.currentThread().interrupt();
+      log.error("Interrupted while downloading or loading LDD for namespace '" + prefix + "' from " + jsonUrl);
+      if (lddInfo.isEmpty()) {
+        log.warn("No previously loaded LDD found for namespace '" + prefix
+            + "'. Fields from this namespace will use 'keyword' data type.");
+      } else {
+        log.warn("Will use previously loaded field definitions for namespace '" + prefix
+            + "' from " + lddInfo.files);
+      }
     } catch (Exception ex) {
       log.error("Failed to download or load LDD for namespace '" + prefix + "' from " + jsonUrl
           + ": " + ExceptionUtils.getMessage(ex));
       if (lddInfo.isEmpty()) {
         log.warn("No previously loaded LDD found for namespace '" + prefix
             + "'. Fields from this namespace will use 'keyword' data type.");
-        return;
       } else {
         log.warn("Will use previously loaded field definitions for namespace '" + prefix
             + "' from " + lddInfo.files);
-        return;
       }
     } finally {
       lddFile.delete();

--- a/src/main/java/gov/nasa/pds/registry/common/es/service/SchemaUpdater.java
+++ b/src/main/java/gov/nasa/pds/registry/common/es/service/SchemaUpdater.java
@@ -137,6 +137,11 @@ public class SchemaUpdater {
     File lddFile;
     try {
       lddFile = File.createTempFile("LDD-", ".JSON");
+      // Restrict permissions to owner only (mitigate publicly writable temp dir risk)
+      lddFile.setReadable(false, false);
+      lddFile.setReadable(true, true);
+      lddFile.setWritable(false, false);
+      lddFile.setWritable(true, true);
     } catch (IOException ex) {
       throw new LddException("Failed to create temp file for LDD download for namespace '"
           + prefix + "': " + ExceptionUtils.getMessage(ex), ex);

--- a/src/main/java/gov/nasa/pds/registry/common/es/service/SchemaUpdater.java
+++ b/src/main/java/gov/nasa/pds/registry/common/es/service/SchemaUpdater.java
@@ -9,6 +9,7 @@ import java.util.Set;
 import org.apache.logging.log4j.LogManager;
 import org.apache.logging.log4j.Logger;
 
+import gov.nasa.pds.registry.common.dd.LddException;
 import gov.nasa.pds.registry.common.dd.LddUtils;
 import gov.nasa.pds.registry.common.util.ExceptionUtils;
 import gov.nasa.pds.registry.common.util.file.FileDownloader;
@@ -79,9 +80,8 @@ public class SchemaUpdater {
         try {
           updateLdd(uri, prefix);
         } catch (Exception ex) {
-          log.error(
-              "Could not update LDD related to schema at URI " + uri + ". Something went wrong.",
-              ex);
+          log.error("Could not update LDD for namespace '" + prefix + "' at URI " + uri
+              + ": " + ex.getMessage() + ". Harvesting will continue with available field definitions.");
         }
       }
     }
@@ -98,7 +98,7 @@ public class SchemaUpdater {
   }
 
 
-  private void updateLdd(String uri, String prefix) throws Exception {
+  private void updateLdd(String uri, String prefix) throws LddException {
     if (uri == null || uri.isEmpty())
       return;
     if (prefix == null || prefix.isEmpty())
@@ -112,12 +112,18 @@ public class SchemaUpdater {
     // Get schema file name
     int idx = jsonUrl.lastIndexOf('/');
     if (idx < 0) {
-      throw new Exception("Invalid schema URI." + uri);
+      throw new IllegalArgumentException("Invalid schema URI: " + uri);
     }
     String schemaFileName = jsonUrl.substring(idx + 1);
 
     // Get stored LDDs info
-    LddVersions lddInfo = ddDao.getLddInfo(prefix);
+    LddVersions lddInfo;
+    try {
+      lddInfo = ddDao.getLddInfo(prefix);
+    } catch (Exception ex) {
+      throw new LddException("Failed to query registry for existing LDD info for namespace '"
+          + prefix + "': " + ExceptionUtils.getMessage(ex), ex);
+    }
 
     // LDD already loaded
     if (lddInfo.files.contains(schemaFileName)) {
@@ -125,19 +131,28 @@ public class SchemaUpdater {
     }
 
     // Download LDD
-    File lddFile = File.createTempFile("LDD-", ".JSON");
+    File lddFile;
+    try {
+      lddFile = File.createTempFile("LDD-", ".JSON");
+    } catch (Exception ex) {
+      throw new LddException("Failed to create temp file for LDD download for namespace '"
+          + prefix + "': " + ExceptionUtils.getMessage(ex), ex);
+    }
 
     try {
       if (fileDownloader.download(jsonUrl, lddFile)) {
         lddLoader.load(lddFile, schemaFileName, prefix);
       }
     } catch (Exception ex) {
-      log.error(ExceptionUtils.getMessage(ex));
+      log.error("Failed to download or load LDD for namespace '" + prefix + "' from " + jsonUrl
+          + ": " + ExceptionUtils.getMessage(ex));
       if (lddInfo.isEmpty()) {
-        log.warn("Will use 'keyword' data type.");
+        log.warn("No previously loaded LDD found for namespace '" + prefix
+            + "'. Fields from this namespace will use 'keyword' data type.");
         return;
       } else {
-        log.warn("Will use field definitions from " + lddInfo.files);
+        log.warn("Will use previously loaded field definitions for namespace '" + prefix
+            + "' from " + lddInfo.files);
         return;
       }
     } finally {
@@ -146,12 +161,11 @@ public class SchemaUpdater {
   }
 
 
-  private String getJsonUrl(String uri) throws Exception {
+  private String getJsonUrl(String uri) {
     if (uri.endsWith(".xsd")) {
-      String jsonUrl = uri.substring(0, uri.length() - 3) + "JSON";
-      return jsonUrl;
+      return uri.substring(0, uri.length() - 3) + "JSON";
     } else {
-      throw new Exception("Invalid schema URI. URI doesn't end with '.xsd': " + uri);
+      throw new IllegalArgumentException("Invalid schema URI - does not end with '.xsd': " + uri);
     }
   }
 

--- a/src/main/java/gov/nasa/pds/registry/common/es/service/SchemaUpdater.java
+++ b/src/main/java/gov/nasa/pds/registry/common/es/service/SchemaUpdater.java
@@ -158,6 +158,7 @@ public class SchemaUpdater {
         log.warn("Will use previously loaded field definitions for namespace '" + prefix
             + "' from " + lddInfo.files);
       }
+      return;
     } catch (Exception ex) {
       log.error("Failed to download or load LDD for namespace '" + prefix + "' from " + jsonUrl
           + ": " + ExceptionUtils.getMessage(ex));

--- a/src/main/java/gov/nasa/pds/registry/common/util/file/FileDownloader.java
+++ b/src/main/java/gov/nasa/pds/registry/common/util/file/FileDownloader.java
@@ -4,7 +4,7 @@ import java.io.File;
 import java.io.FileOutputStream;
 import java.io.IOException;
 import java.io.InputStream;
-import java.util.ArrayList;
+import java.util.concurrent.CopyOnWriteArrayList;
 import javax.net.ssl.SSLContext;
 import org.apache.http.HttpEntity;
 import org.apache.http.StatusLine;
@@ -39,8 +39,8 @@ import gov.nasa.pds.registry.common.util.CloseUtils;
  */
 public class FileDownloader
 {
-  final private static ArrayList<String> ignore = new ArrayList<String>();
-  final private static ArrayList<String> failed = new ArrayList<String>();
+  final private static CopyOnWriteArrayList<String> ignore = new CopyOnWriteArrayList<String>();
+  final private static CopyOnWriteArrayList<String> failed = new CopyOnWriteArrayList<String>();
     private Logger log;
     private int numRetries = 3;
     

--- a/src/main/java/gov/nasa/pds/registry/common/util/file/FileDownloader.java
+++ b/src/main/java/gov/nasa/pds/registry/common/util/file/FileDownloader.java
@@ -2,6 +2,7 @@ package gov.nasa.pds.registry.common.util.file;
 
 import java.io.File;
 import java.io.FileOutputStream;
+import java.io.IOException;
 import java.io.InputStream;
 import java.util.ArrayList;
 import javax.net.ssl.SSLContext;
@@ -39,6 +40,7 @@ import gov.nasa.pds.registry.common.util.CloseUtils;
 public class FileDownloader
 {
   final private static ArrayList<String> ignore = new ArrayList<String>();
+  final private static ArrayList<String> failed = new ArrayList<String>();
     private Logger log;
     private int numRetries = 3;
     
@@ -64,10 +66,15 @@ public class FileDownloader
      * @param toFile Save to this file
      * @throws Exception an exception
      */
-    public boolean download(String fromUrl, File toFile) throws Exception
+    public boolean download(String fromUrl, File toFile) throws IOException, InterruptedException
     {
+        if(failed.contains(fromUrl))
+        {
+            throw new IOException("Could not download " + fromUrl + " (already failed in this running instance, will not retry).");
+        }
+
         int count = 0;
-        
+
         while(!ignore.contains(fromUrl))
         {
             try
@@ -86,8 +93,8 @@ public class FileDownloader
                 }
                 else
                 {
-                  ignore.add(fromUrl);
-                  throw new Exception("Could not download " + fromUrl + " and will not try again in this running instance.");
+                  failed.add(fromUrl);
+                  throw new IOException("Could not download " + fromUrl + " and will not try again in this running instance.");
                 }
             }
         }

